common/Scripts/ModifyWindowsOptionalFeatures.cs

```diff
@@ -0,0 +1,227 @@
+
+        return exitCode;
+    }
+
+    public enum ExitCode
+    {
+        Success = 0,
+        NoChange = 1,
+        Failure = 2,
+        Cancelled = 3,
+    }
+
+    private static ExitCode FromExitCode(int exitCode)
+    {
+        return exitCode switch
+        {
+            0 => ExitCode.Success,
+            1 => ExitCode.NoChange,
+            2 => ExitCode.Failure,
+            _ => ExitCode.Cancelled,
+        };
+    }
+
+    /// <summary>
+    /// PowerShell script for modifying Windows optional features.
+    ///
+    /// This script takes a string argument representing feature names and their desired states (enabled or disabled).
+    /// It parses this string into a dictionary, iterates over each feature, and performs the necessary enable or disable operation based on the desired state.
+    ///
+    /// The script defines the following possible exit statuses:
+    /// - OperationSucceeded (0): All operations (enable or disable) succeeded and no restart is needed.
+    /// - OperationSkipped (1): No operations were performed because the current state of all features matched the desired state.
+    /// - OperationFailed (2): At least one operation failed.
+    ///
+    /// Only features present in "validFeatures" are considered valid. If an invalid feature name is encountered, the script exits with OperationFailed.
+    /// This list should be kept consistent with the list of features in the WindowsOptionalFeatureNames class.
+    ///
+    /// </summary>
+    private const string Script =
+@"
+enum OperationStatus
+{
+    OperationSucceeded = 0
+    OperationSkipped = 1
+    OperationFailed = 2
+}
+
+$validFeatures = @(
+    'Containers',
+    'HostGuardian',
+    'Microsoft-Hyper-V-All',
+    'Microsoft-Hyper-V-Tools-All',
+    'Microsoft-Hyper-V',
+    'VirtualMachinePlatform',
+    'HypervisorPlatform',
+    'Containers-DisposableClientVM',
+    'Microsoft-Windows-Subsystem-Linux'
+)
+
+function ModifyFeatures($featuresString)
+{
+    $features = ConvertFrom-StringData $featuresString
+    $anyOperationFailed = $false
+    $anyOperationPerformed = $false
+
+    foreach ($feature in $features.GetEnumerator())
+    {
+        $featureName = $feature.Key
+        if ($featureName -notin $validFeatures)
+        {
+            exit [OperationStatus]::OperationFailed
+        }
+
+        $isEnabled = [bool]::Parse($feature.Value);
+        $featureState = Get-WindowsOptionalFeature -FeatureName $featureName -Online | Select-Object -ExpandProperty State;
+        $currentEnabled = $featureState -eq 'Enabled';
+
+        if ($currentEnabled -ne $isEnabled)
+        {
+            $anyOperationPerformed = $true
+            if ($isEnabled)
+            {
+                $enableResult = Enable-WindowsOptionalFeature -Online -FeatureName $featureName -All -NoRestart
+                if ($enableResult -eq $null)
+                {
+                    $anyOperationFailed = $true
+                }

```

Comment:
i don't think this works.

say $enableResult is set to a valid object because the first WOF succeeded,

if then on the 2nd loop enabling the WOF fails, your $enableResult is still valid and your code will treat it as succeeded.

(powershell variable scoping doesn't really exist.)

Reply:
i would use the if(!$?) idiom to see if Enable-WOF failed

Reply:
you're right, I meant to address this when you brought it up before and forgot. Thinking about it more now, I think just bailing out early if anything fails is probably best rather than trying this best effort nonsense.